include/asioexec/completion_token.hpp

```diff
@@ -0,0 +1,411 @@
+      h_ = nullptr;
+      ::stdexec::set_error(std::move(r_), std::current_exception());
+    }
+    if (prev) {
+      *prev = completed;
+    }
+    if (!completed) {
+      completed_ = prev;
+    }
+  }
+protected:
+  std::optional<
+    ::stdexec::stop_callback_for_t<
+      ::stdexec::stop_token_of_t<
+        ::stdexec::env_of_t<Receiver>>,
+    stop_callback>> callback_;
+};
+
+template<typename Receiver>
+class completion_handler {
+  operation_state_base<Receiver>* self_;
+  template<typename T, typename... Args>
+  void complete_(const T& channel, Args&&... args) noexcept {
+    assert(self_);
+    auto&& r = self_->r_;
+    self_->h_ = nullptr;
+    if (self_->completed_) {
+      assert(!*self_->completed_);
+      *self_->completed_ = true;
+      self_->completed_ = nullptr;
+    }
+    self_ = nullptr;
+    std::invoke(channel, std::move(r), std::forward<Args>(args)...);
+  }
+public:
+  constexpr explicit completion_handler(operation_state_base<Receiver>& self)
+    noexcept : self_(&self)
+  {
+    assert(!self_->h_);
+    self_->h_ = this;
+  }
+  constexpr completion_handler(completion_handler&& other) noexcept
+    : self_(std::exchange(other.self_, nullptr))
+  {
+    if (self_) {
+      self_->h_ = this;
+    }
+  }
+  completion_handler& operator=(const completion_handler&) = delete;
+  constexpr ~completion_handler() noexcept {
+    if (self_ && !self_->completed_) {
+      ::stdexec::set_stopped(std::move(self_->r_));
+    }
+  }
+  template<typename... Args>
+  void operator()(Args&&... args) && noexcept {
+    complete_(::stdexec::set_value, std::forward<Args>(args)...);
+  }
+  using cancellation_slot_type = ::boost::asio::cancellation_slot;
+  auto get_cancellation_slot() const noexcept {
+    assert(self_);
+    return self_->signal_.slot();
+  }
+  constexpr operation_state_base<Receiver>& state() const noexcept {
+    assert(self_);
+    return *self_;
+  }
+  constexpr void release() noexcept {
+    self_ = nullptr;
+  }
+};
+
+template<typename Receiver, typename Initiation, typename Args>
+class operation_state : operation_state_base<Receiver> {
+  using base_ = operation_state_base<Receiver>;
+  Initiation init_;
+  Args args_;
+public:
+  template<typename T, typename U, typename V>
+    requires
+      std::constructible_from<base_, T> &&
+      std::constructible_from<Initiation, U> &&
+      std::constructible_from<Args, V>
+  constexpr explicit operation_state(T&& t, U&& u, V&& v) noexcept(
+    std::is_nothrow_constructible_v<base_, T> &&
+    std::is_nothrow_constructible_v<Initiation, U> &&
+    std::is_nothrow_constructible_v<Args, V>)
+    : base_(std::forward<T>(t)),
+      init_(std::forward<U>(u)),
+      args_(std::forward<Args>(v))
+  {}
+  constexpr void start() & noexcept {
+    try {
+      std::apply(
+        [&](auto&&... args) {
+          std::invoke(
+            std::move(init_),
+            completion_handler<Receiver>(*this),
+            std::forward<decltype(args)>(args)...);
```

Comment:
if we are comfortable saying that `init_` is never going to be a `completion_handler` member pointer, then we can change this to not use `std::invoke` also.

Reply:
I don't know that we can say that? I'd need to read very finely the expectations that `async_initiate` has for the `Initiation`.